integration-tests/tests/src/tests/sync/flexible.ts

```diff
@@ -2075,4 +2081,400 @@ describe("Flexible sync", function () {
+        Realm.ProgressDirection.Download,
+        Realm.ProgressMode.ForCurrentlyOutstandingWork,
+        callback,
+      );
+    });
+
+    it("old callback style is not allowed", async function (this: RealmContext) {
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/no-empty-function
+      const callback = spy((transferable: number, transferred: number) => {});
+      expect(() => {
+        this.realm.syncSession?.addProgressNotification(
+          Realm.ProgressDirection.Download,
+          Realm.ProgressMode.ForCurrentlyOutstandingWork,
+          callback,
+        );
+      }).to.throw();
+
+      expect(callback).not.called;
+    });
+
+    describe("with ProgressMode.ReportIndefinitely", function () {
+      describe(`with ProgressDirection.Upload`, function () {
+        it("should not call callback when there is nothing to upload", async function (this: RealmContext) {
+          const realm = this.realm;
+          // eslint-disable-next-line @typescript-eslint/no-empty-function, @typescript-eslint/no-unused-vars
+          const callback = spy((estimate: number) => {});
+
+          realm.syncSession?.addProgressNotification(
+            Realm.ProgressDirection.Upload,
+            ProgressMode.ReportIndefinitely,
+            callback,
+          );
+
+          expect(callback).not.called;
+        });
+
+        it("should be called multiple times with different values during uploads", async function (this: RealmContext) {
+          const realm = this.realm;
+          // eslint-disable-next-line @typescript-eslint/no-empty-function, @typescript-eslint/no-unused-vars
+          const callback = spy((estimate: number) => {});
+
+          realm.syncSession?.addProgressNotification(
+            Realm.ProgressDirection.Upload,
+            Realm.ProgressMode.ReportIndefinitely,
+            callback,
+          );
+
+          await this.realm.syncSession?.uploadAllLocalChanges();
+          await this.realm.syncSession?.downloadAllServerChanges();
+
+          // TODO: This callback should not be called at this stage but seems flakey
+          // and gets called with 1.0 at times, likely because of a race condition.
+          expect(callback.notCalled || callback.calledWith(1.0)).is.true;
+          const subscription = await realm.objects(Person).subscribe();
+
+          realm.write(() => {
+            realm.create(Person, {
+              _id: new BSON.ObjectId(),
+              name: "Heavy",
+              age: 36,
+              data: new ArrayBuffer(1_000_000),
+            });
+          });
+
+          await realm.syncSession?.uploadAllLocalChanges();
+          subscription.unsubscribe();
+
+          // TODO: Find a way to fix
+          // There should be at least one point where the progress is not yet finished.
+          // expect(callback.args.find(([estimate]) => estimate < 1)).to.not.be.undefined;
+
+          expect(callback.withArgs(1.0)).called;
```

Comment:
Unsure if `withArgs()` is the correct usage here since that seems to create a new spy (according to their API docs). If using that, you probably want to put the `called` onto the spy instead (similar to the above where `callback.notCalled` is used):
```js
expect(callback.withArgs(1.0).called).to.be.true
```

However, there are also these APIs:
```js
callback.calledWith(..)
callback.calledWithExactly(..)
```